src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Cohost/OpenDocumentGenerator.cs

```diff
@@ -0,0 +1,148 @@
+                // for a document before we get the notifications about it being added to any projects. Lets try refreshing before
+                // we worry.
+                _logger.LogDebug("[Cohost] Refreshing virtual documents, and waiting for them, (for {hostDocumentUri})", documentSnapshot.Uri);
+
+                var task = _csharpVirtualDocumentAddListener.WaitForDocumentAddAsync(cancellationToken);
+                _documentManager.RefreshVirtualDocuments();
+                _ = await task.ConfigureAwait(true);
+
+                // Since we're dealing with snapshots, we have to get the new ones after refreshing
+                if (!_documentManager.TryGetDocument(documentSnapshot.Uri, out var newDocumentSnapshot) ||
+                    !newDocumentSnapshot.TryGetAllVirtualDocuments<CSharpVirtualDocumentSnapshot>(out virtualDocuments))
+                {
+                    // This should never happen.
+                    // The server clearly wants to tell us about a document in a project, but we don't know which project it's in.
+                    // Sadly there isn't anything we can do here to, we're just in a state where the server and client are out of
+                    // sync with their understanding of the document contents, and since changes come in as a list of changes,
+                    // the user experience is broken. All we can do is hope the user closes and re-opens the document.
+                    Debug.Fail($"Server wants to update {documentSnapshot.Uri} in {projectKey} but we don't know about the document being in any projects");
+                    _logger.LogError("[Cohost] Server wants to update {hostDocumentUri} in {projectKeyId} by we only know about that document in misc files. Server and client are now out of sync.", documentSnapshot.Uri, projectKey);
+                    return null;
+                }
+            }
+
+            foreach (var virtualDocument in virtualDocuments)
+            {
+                if (virtualDocument.ProjectKey.Equals(projectKey))
+                {
+                    _logger.LogDebug("[Cohost] Found C# virtual doc for {version}: {uri}", version, virtualDocument.Uri);
+
+                    return virtualDocument;
+                }
+            }
+
+            if (virtualDocuments.Length > 1)
+            {
+                // If the particular document supports multiple virtual documents, we don't want to try to update a single one
+                // TODO: Remove this eventually, as it is a possibly valid state (see comment below) but this assert will help track
+                // down bugs for now.
+                Debug.Fail("Multiple virtual documents seem to be supported, but none were updated, which is not impossible, but surprising.");
+            }
+
+            _logger.LogDebug("[Cohost] Couldn't find any virtual docs for {version} of {uri}", version, documentSnapshot.Uri);
```

Comment:
>"[Cohost] Couldn't find any virtual docs for {version} of {uri}" [](http://example.com/codeflow?start=29&length=64)

nit: should we also add "with project key {projectKey}" since this will get logged if we found one or more virtualDocuments but none had the desired project key?